# 03 · Reversible, same batch: 50M tokens, batch 32

The rule chosen in notebook 01, at exactly the baseline's batch, data order, schedule and seed. The only
difference from notebook 02 is how the residual stream moves between layers and what the backward pass
keeps. This is the like-for-like comparison: same work, what does reversibility cost in time and save in
memory?

In [1]:
import sys, os, json, pathlib, subprocess
IN_COLAB = "google.colab" in sys.modules
if IN_COLAB and not pathlib.Path("ERA_V5").exists():
    subprocess.run(["git", "clone", "--depth", "1", "https://github.com/amukka/ERA_V5.git"], check=True)
    os.chdir("ERA_V5/session13_Distributed_Training")
ROOT = pathlib.Path.cwd()
while not (ROOT / "src").exists() and ROOT != ROOT.parent:
    ROOT = ROOT.parent
os.chdir(ROOT); sys.path.insert(0, str(ROOT))
if not (ROOT / "data" / "train.bin").exists():          # ~5 minutes: stream FineWeb-Edu, train the BPE
    subprocess.run([sys.executable, "tools/prepare_data.py"], check=True)

import torch, platform
from src import experiments as X
from src.train import pick_device
dev = pick_device()
print("device:", dev, "| torch", torch.__version__, "| python", platform.python_version(), "|", platform.platform())
if dev == "mps":
    print(f"MPS recommended working set: {torch.mps.recommended_max_memory()/2**30:.2f} GiB")
if dev == "cuda":
    print(torch.cuda.get_device_name(), f"{torch.cuda.get_device_properties(0).total_memory/2**30:.1f} GiB")
print(json.dumps(json.loads((ROOT / "data" / "meta.json").read_text()), indent=1))

device: mps | torch 2.13.0 | python 3.14.5 | macOS-26.6.2-arm64-arm-64bit-Mach-O
MPS recommended working set: 11.84 GiB
{
 "source": "HuggingFaceFW/fineweb-edu, config sample-10BT, streamed in order",
 "vocab_size": 8192,
 "eos_id": 0,
 "train_tokens": 56000000,
 "train_docs": 46080,
 "val_tokens": 1000000,
 "val_docs": 1024,
 "chars_per_token": 3.886,
 "note": "val.bin comes from the tokenizer's training documents; train.bin from the documents after them",
 "seconds": 314.3
}


**The model** (identical in every run): GPT, 10 layers, d_model 384, 6 heads, MLP 4x, sequence 512,
byte-level BPE vocabulary 8,192 with the head tied to the embedding. **21,053,184 parameters**, of which
17,710,848 are outside the embedding tables. No dropout anywhere, because a reversible backward pass
recomputes each block and a random mask would recompute differently. AdamW (0.9, 0.95, wd 0.1), peak LR
1e-3, 3% warmup then cosine to 10%, gradient clip 1.0, fp32.

In [2]:
chosen = json.loads((X.RESULTS / "chosen_variant.json").read_text())
print("rule:", chosen["variant"], "h =", chosen["h"])
rev = X.train(name=f"reversible_{chosen['variant']}_b32", variant=chosen["variant"], h=chosen["h"],
              batch=X.BASE_BATCH, tokens=X.TOKENS, eval_every=250, log_every=10)

rule: momentum h = 0.5


{"name": "reversible_momentum_b32", "device": "mps", "config": {"vocab_size": 8192, "seq_len": 512, "n_layer": 10, "n_head": 6, "d_model": 384, "variant": "momentum", "h": 0.5, "fp_iters": 6, "extra": {}}, "params": 21053184, "params_non_embedding": 17710848, "batch": 32, "tokens_per_step": 16384, "steps": 3051, "peak_lr": 0.001, "warmup": 91, "amp_bf16": 0, "torch": "2.13.0", "machine": "arm64", "processor": "arm"}


step     0/3051  tok   0.02M  loss 9.0825  gnorm 3.503  0 tok/s  0.0 min


memory audit at step 3: exact peak 2,755 MiB in aten._softmax_backward_data.default


step   100/3051  tok   1.64M  loss 6.2163  gnorm 0.401  8,947 tok/s  3.0 min


step   200/3051  tok   3.28M  loss 5.8474  gnorm 0.431  8,929 tok/s  6.1 min


            eval @ 4.08M tokens: val loss 5.7371


step   300/3051  tok   4.92M  loss 5.6471  gnorm 0.424  8,932 tok/s  9.1 min


step   400/3051  tok   6.55M  loss 5.4108  gnorm 0.434  8,910 tok/s  12.2 min


            eval @ 8.18M tokens: val loss 5.3129


step   500/3051  tok   8.19M  loss 5.2184  gnorm 0.593  9,093 tok/s  15.3 min


step   600/3051  tok   9.83M  loss 5.0354  gnorm 0.500  8,932 tok/s  18.4 min


step   700/3051  tok  11.47M  loss 4.9698  gnorm 0.529  8,940 tok/s  21.4 min


            eval @ 12.27M tokens: val loss 5.0209


step   800/3051  tok  13.11M  loss 4.8879  gnorm 0.540  8,920 tok/s  24.5 min


step   900/3051  tok  14.75M  loss 4.8246  gnorm 0.549  8,930 tok/s  27.5 min


            eval @ 16.37M tokens: val loss 4.8049


step  1000/3051  tok  16.38M  loss 4.7782  gnorm 0.549  9,083 tok/s  30.6 min


step  1100/3051  tok  18.02M  loss 4.7018  gnorm 0.552  8,919 tok/s  33.6 min


step  1200/3051  tok  19.66M  loss 4.6362  gnorm 0.584  8,914 tok/s  36.7 min


            eval @ 20.46M tokens: val loss 4.6270


step  1300/3051  tok  21.30M  loss 4.6164  gnorm 0.558  8,920 tok/s  39.8 min


step  1400/3051  tok  22.94M  loss 4.5592  gnorm 0.566  8,921 tok/s  42.8 min


            eval @ 24.56M tokens: val loss 4.4800


step  1500/3051  tok  24.58M  loss 4.5298  gnorm 0.649  9,081 tok/s  45.9 min


step  1600/3051  tok  26.21M  loss 4.4049  gnorm 0.614  8,901 tok/s  48.9 min


step  1700/3051  tok  27.85M  loss 4.3776  gnorm 0.658  8,906 tok/s  52.0 min


            eval @ 28.66M tokens: val loss 4.3486


step  1800/3051  tok  29.49M  loss 4.3358  gnorm 0.674  8,915 tok/s  55.1 min


step  1900/3051  tok  31.13M  loss 4.3240  gnorm 0.642  8,917 tok/s  58.1 min


            eval @ 32.75M tokens: val loss 4.2388


step  2000/3051  tok  32.77M  loss 4.1860  gnorm 0.725  9,072 tok/s  61.2 min


step  2100/3051  tok  34.41M  loss 4.1828  gnorm 0.652  8,914 tok/s  64.3 min


step  2200/3051  tok  36.04M  loss 4.1363  gnorm 0.701  8,904 tok/s  67.3 min


            eval @ 36.85M tokens: val loss 4.1560


step  2300/3051  tok  37.68M  loss 4.1525  gnorm 0.710  8,924 tok/s  70.4 min


step  2400/3051  tok  39.32M  loss 4.0679  gnorm 0.689  8,926 tok/s  73.4 min


            eval @ 40.94M tokens: val loss 4.0963


step  2500/3051  tok  40.96M  loss 4.0680  gnorm 0.744  9,086 tok/s  76.5 min


step  2600/3051  tok  42.60M  loss 4.1049  gnorm 0.725  8,914 tok/s  79.6 min


step  2700/3051  tok  44.24M  loss 4.0332  gnorm 0.756  8,913 tok/s  82.6 min


            eval @ 45.04M tokens: val loss 4.0526


step  2800/3051  tok  45.88M  loss 4.0579  gnorm 0.718  8,932 tok/s  85.7 min


step  2900/3051  tok  47.51M  loss 3.9301  gnorm 0.682  8,924 tok/s  88.7 min


            eval @ 49.14M tokens: val loss 4.0264


step  3000/3051  tok  49.15M  loss 4.0640  gnorm 0.747  9,086 tok/s  91.8 min


step  3050/3051  tok  49.97M  loss 4.0007  gnorm 0.704  8,923 tok/s  93.3 min


            eval @ 49.97M tokens: val loss 4.0206


reversibility at final weights: grad rel err 9.58e-05, cosine 1.000000, state rebuild err 2.89e-05
{"tok_per_s": 8924.7, "final_train_loss": 4.000720977783203, "final_val_loss": 4.020622986555099, "peak_exact_mib": 2754.9, "peak_alloc_mib": 1620.7, "peak_driver_mib": 6211.1, "train_seconds": 5599.2}


In [3]:
keys = ["tokens_trained", "steps", "final_train_loss", "final_train_loss_avg_last5pct", "final_val_loss",
        "tok_per_s", "train_seconds", "peak_exact_mib", "peak_exact_op", "peak_alloc_mib", "peak_driver_mib"]
print(json.dumps({k: rev[k] for k in keys}, indent=1))
print(json.dumps(rev["reversibility"], indent=1))

{
 "tokens_trained": 49971200,
 "steps": 3051,
 "final_train_loss": 4.000720977783203,
 "final_train_loss_avg_last5pct": 4.01357962290446,
 "final_val_loss": 4.020622986555099,
 "tok_per_s": 8924.7,
 "train_seconds": 5599.2,
 "peak_exact_mib": 2754.9,
 "peak_exact_op": "aten._softmax_backward_data.default",
 "peak_alloc_mib": 1620.7,
 "peak_driver_mib": 6211.1
}
{
 "init": {
  "grad_rel_err": 2.200800561903834e-06,
  "grad_cosine": 0.9999999999972712,
  "state_rebuild_rel_err_per_layer": [
   5.324042831489351e-06,
   2.7300793590256944e-06,
   1.537154957986786e-06,
   9.912050700222608e-07,
   6.691512908219011e-07,
   5.152484163772897e-07,
   4.32310315545692e-07,
   2.8608965862986224e-07,
   2.3783076130712288e-07
  ],
  "state_rebuild_rel_err_max": 5.324042831489351e-06
 },
 "final": {
  "grad_rel_err": 9.583595126249242e-05,
  "grad_cosine": 0.999999995413234,
  "state_rebuild_rel_err_per_layer": [
   2.8914064387208782e-05,
   6.350663625198649e-06,
   4.238381279719761e-06,
 